DirectProgramming/DPC++FPGA/include/cached_local_memory.hpp

```diff
@@ -0,0 +1,123 @@
+#ifndef __CACHED_LOCAL_MEMORY_HPP__
+#define __CACHED_LOCAL_MEMORY_HPP__
+
+#include <sycl/ext/intel/ac_types/ac_int.hpp>
+
+#include "constexpr_math.hpp"   // DirectProgramming/DPC++FPGA/include
+
+namespace fpga_tools {
+
+template <typename T,             // type to store in the memory
+          size_t k_mem_depth,     // depth of the memory
+          size_t k_cache_depth    // number of elements in the cache
+         >
+class CachedLocalMemory {
+ public:
+  static constexpr int kNumAddrBits = fpga_tools::CeilLog2(k_mem_depth);
+  using addr_t = ac_int<kNumAddrBits, false>;
+
+  CachedLocalMemory() {}
+  CachedLocalMemory(T init_val) { 
+    for (int i = 0; i < k_mem_depth; i++) {
+      write(i, init_val);
+    }
+  }
+  
+  CachedLocalMemory(const CachedLocalMemory&) = delete;
+  CachedLocalMemory& operator=(const CachedLocalMemory&) = delete;
+
+  // explicitly communicate to developers that we don't want to support a
+  // square bracket operator that returns a reference, as it would allow
+  // modification of the memory without updating the cache
+  template <typename I>
+  T& operator[](I addr) = delete;
+
+  // we can support the square bracket operator that returns a const ref
+  const T& operator[](addr_t addr) const { return read(addr); }
+
+  void write(addr_t addr, T val) {
+    // Shift the values in the cache
+    #pragma unroll
+    for (int i = 0; i < k_cache_depth-1; i++) {
+      cache_val_[i] = cache_val_[i + 1];
+      cache_addr_[i] = cache_addr_[i + 1];
+    }
+
+    // write the new value into the array and the cache
+    data_[addr] = val;
+    cache_val_[k_cache_depth-1] = val;
+    cache_addr_[k_cache_depth-1] = addr;
+  }
+
+  T read(addr_t addr) { 
+    // Get the value from the local memory
+    auto val = data_[addr];
+
+    // If this address was recently written, take the value from the cache
+    #pragma unroll
+    for (int i = 0; i < k_cache_depth; i++) {
+      if (cache_addr_[i] == addr) val = cache_val_[i];
+    }
+
+    return val;
+  }
```

Comment:
I'm not sure to understand how this is working as a cache (but I can see that this was also what was done in the original cached_local_memory.hpp file).
When I think about a cache, I think about a "controler" that is the only interface with the memory. So no direct interaction with the memory is possible.
The cache will however manage:
- the writes to the memory if there is a write operation and it is full
- the reads from the memory if there is a read operation and the data is not present in the cache.

Here what is implemented is an history of the last writes - but I'm guessing that it also qualifies as a cache?

Now I'm (positively) surprised that our compiler is able to take advantage of this construct!
Can you confirm that the hardware generated is able to bypass the memory read (specially its latency) to read directly from the cache if the data is available?

Reply:
The cache stores the most recent N writes to the memory.  If you read from the memory, and the read address matches an address in the cache, you pull data from the cache instead of from the memory itself.  That means as long as the read latency of the memory is <=N, you will get the correct value: values written more than N cycles ago can be safely fetched from the memory, and values written in the last N cycles are fetched from the cache, since they can't be read safely from the memory.  Does that make sense?

Reply:
And the memory read still happens -- but the result of the read is discarded if there's a cache hit.

Reply:
It makes perfect sense, I'm just happily surprised that the compiler does not insert pipeline registers after the cache read to match the memory read latency.